Added Gallery Hijack

Ristellise opened the pull request:
Gradio generates it's own temp folders for gallery component and on Windows, does not clear them after the server has been shut down, which potentially can lead to storage issues and potentially some privacy issues if you think hard enough about it.
This hijack serves to override Gradio's Gallery component with our own Gallery component so as to have control on temporary folder.

Notes:
 - The class name is important to trick Gradio's CSS UI (Svelte) into thinking it's the original class instead of our hijacked one.
 - `getcwd()` returns the root folder if the uses runs the `.bat` file. We use a hardcoded `gradio_temp` folder for now
 - the first `shutil.rmtree(self.temp_dir)` is to remove the temp folder created by the first call.
 - Once the server starts, the `gradio_temp` folder is wiped.

Other considerations:
 - The temp folder can be helpful to recover the generated images if deleted from txt2img and various other folders. But I believe not much folks out there do know about it.
 - I haven't done much testing on it but it runs soo...

Ristellise committed:
Add Gallery Hijack

Ristellise committed:
Fix img2img button

36DB commented:
Can you make this feature as a toggle option in the settings? As you mentioned some people might want to keep the temp images as a backup plan.

AUTOMATIC1111 commented:
`os.path.join(os.getcwd(), "gradio_temp")` seems really wrong. There's no way this is where gradio writes by default.

Ristellise committed:
Added User Configurability.

Ristellise committed:
Tweaked Config message, Bugfix Gallery.
 - Turns out The Web UI does not like it if you put a class behind a conditional statement.

Ristellise commented:
> `os.path.join(os.getcwd(), "gradio_temp")` seems really wrong. There's no way this is where gradio writes by default.

It write to `C:\Users\<Username>\AppData\Local\Temp\tmp<randomcharacters>` on Windows. (Or the path returned by [`gettempdir()`](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir) + "tmp<RandomString>"). The `os.path.join(...)` is used to override the temp folder for the gallery. I do agree this a not too ideal option. As such, I've changed it to accept user input and resolving the final path with `os.path.realpath()`

> Can you make this feature as a toggle option in the settings? As you mentioned some people might want to keep the temp images as a backup plan.

Done this in the latest commit. The config option `outdir_gradio_temp` is located in the `Paths for saving` section. :)

Ristellise commented:
I just realized that to add in the option for user configurable temp path breaks `image_from_url_text` since it relies on a tempfile provided by python. if we could pass `opts` from UI to `generation_parameters_copypaste` that would fix the issue but I'm not sure how to achieve it.

Another solution would be to override [`tempdir`](https://docs.python.org/3/library/tempfile.html#tempfile.tempdir) route but I'm not sure how much would break.

Ristellise committed:
Merge branch 'master' of https://github.com/AUTOMATIC1111/stable-diffusion-webui

Ristellise committed:
Bugfix Temp dir

Ristellise committed:
Merge branch 'master' of https://github.com/AUTOMATIC1111/stable-diffusion-webui

Ristellise commented:
Thought of this implementation, any better implementations are appreciated.

Ristellise commented:
Implemented in: https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/b006382784a2f0887317bb60ea49d19b50a5dc7e. As such this PR is made not as useful. Closing.
